### PR TITLE
revert changes to the destruction of systray from the mainframe

### LINKED ISSUE #12238: NVDA Alpha fails to clean its system tray icon after restart
### Steps to reproduce:
Start NVDA Alpha, restart it ten times in a row and inspect system tray.
### Actual behavior:
There are 10 NVDA icons in system tray - 9 of them are dead and would disappear when clicked.
### Expected behavior:
There should be just one icon for the currently running instance of NVDA.

#### NVDA installed/portable/running from source:
Portable
#### NVDA version:
NVDA version alpha-22160,656ecb5f
#### Windows version:
Windows 7 X64
#### Name and version of other software in use when reproducing the issue:
None
#### Does the issue still occur after restarting your computer?
Yes
#### Have you tried any other versions of NVDA? If so, please report their behaviors.
2020.4 works correctly some older Alphas works also.

#### If add-ons are disabled, is your problem still occurring?
Yes
#### Did you try to run the COM registry fixing tool in NVDA menu / tools?
NO

cc @seanbudd

### PR DESCRIPTION
**Pull request by seanbudd**
### Link to issue number:

Closes #12238 

### Summary of the issue:

System tray icons are not cleaned properly when exiting/restarting NVDA. This is due to an incorrect assumption in #12183 that the manual destruction of the system tray icon was no longer necessary. 

### Description of how this pull request fixes the issue:

Revert the changes in #12183 that manually destroyed the system tray icon and add logging to the process. 

Garbage collection is not enough with wx; Destroy() must eventually be called on all objects. wx does handle destruction of children, but the system tray icon and its menu are not children, so we need to handle them explicitly. https://github.com/nvaccess/nvda/commit/8ab9d48527f456d8497bc02725ffff987a33d715

### Testing strategy:

Manually test the steps in #12238

Ensure the silent installation process doesn't crash (ie a regression on what #12183 fixed for #12153)

### Known issues with pull request:

None

### Change log entry:

None, fixes regression

### Code Review Checklist:

- [ ] Pull Request description is up to date.
- [x] Unit tests.
- [x] System (end to end) tests.
- [x] Manual tests.
- [x] User Documentation.
- [x] Change log entry.
- [x] Context sensitive help for GUI changes.

### CODE REVIEW

**Comment by feerrenrut** [comment]
It would be good if we had a handle on when it was necessary to manually call destroy on children in wx. Have you seen this in the docs?

**Comment by feerrenrut** [comment]
So the [Window deletion overiview](https://wxpython.org/Phoenix/docs/html/window_deletion_overview.html) says:
>Child windows are deleted from within the parent destructor. This includes any children that are themselves frames or dialogs, so you may wish to close these child frame or dialog windows explicitly from within the parent close handler.

**Comment by seanbudd** [comment]
> It would be good if we had a handle on when it was necessary to manually call destroy on children in wx. Have you seen this in the docs?

It isn't documented anywhere officially as far as I can tell, just forums/blogs like the following: 

- http://www.blog.pythonlibrary.org/2011/12/13/wxpython-101-creating-taskbar-icons/
- https://discuss.wxpython.org/t/problem-with-closing-the-app-using-taskbaricon/24962/8

The original commit which introduced this had a good message https://github.com/nvaccess/nvda/commit/8ab9d48527f456d8497bc02725ffff987a33d715:

>Garbage collection is not enough with wx; Destroy() must be called on all objects. wx does handle destruction of children, but the system tray icon and its menu are not children, so we need to handle them explicitly. The application cannot terminate if objects are not destroyed. This fixes the issue where NVDA refuses to exit. There are still many cases in settingsDialogs that don't call Destroy() that will cause this bug now that they are not children of the main frame, so these need to be fixed also.

The cases mentioned surrounding Dialogs should be fixed with #12183

**Comment by feerrenrut** [comment]
Right, thanks for looking into this. It seems the key is:

> wx does handle destruction of children, but the system tray icon and its menu are not children,

Would it make sense to ensure that they are children?

**Comment by seanbudd** [comment]
`wx.adv.TaskBarIcon` can't be a child of our main frame window as it [doesn't inherit from](https://wxpython.org/Phoenix/docs/html/wx.adv.TaskBarIcon.html) `wx.Window` and the children of Windows [must inherit](https://wxpython.org/Phoenix/docs/html/wx.Window.html#wx.Window.GetChildren) from `wx.Window`. This is the same reason why we can only call `Destroy`. 

This is also all the same for our taskbar menu `wx.Menu` https://wxpython.org/Phoenix/docs/html/wx.Menu.html

**Comment by feerrenrut** [comment]
Ah, good find. In this case, perhaps we can add these to a list of "must be explicitly destroyed on exit" controls on the App of somewhere appropriate and have it iterated over calling destroy? I'm thinking that if we call this out on the app that there are certain types of controls that need special handling, how to differentiate them from the others, and a mechanism for destroying them on shutdown it will be easier to understand in the future.

**Comment by seanbudd** [comment]
I think the reason why they were being destroyed from where they are currently is because the same class is initialising them.

I could move it all to `safeAppExit`, but you can see the dilemma of where this code belongs.

File: source/gui/__init__.py [hunk 1]
```diff
@@ -53,6 +53,8 @@ def __init__(self):
 		style = wx.DEFAULT_FRAME_STYLE ^ wx.MAXIMIZE_BOX ^ wx.MINIMIZE_BOX | wx.FRAME_NO_TASKBAR
 		super(MainFrame, self).__init__(None, wx.ID_ANY, versionInfo.name, size=(1,1), style=style)
 		self.Bind(wx.EVT_CLOSE, self.onExitCommand)
+
+		# Ensure that this icon is destroyed when this is destroyed as wx doesn't track it (#12243)
```
Comment (feerrenrut) [inline source/gui/__init__.py:57]: I think this should just duplicated the comment in the Destroy method, it's helpful to know why it won't be automatically destroyed.

File: source/gui/__init__.py [hunk 2]
```diff
@@ -75,6 +77,13 @@ def __init__(self):
 				self.Show()
 				self.Hide()
 
+	def Destroy(self):
```
Comment (feerrenrut) [inline source/gui/__init__.py:80]: Note, if wx calls destroy automatically for some reason this won't be called (see https://github.com/wxWidgets/Phoenix/issues/630). This has tripped us up in the past. It might be better to move the `sysTrayIcon` destruction into an `onDestroy` event handler and register for that event.

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/gui/__init__.py (hunk 2 above)
Comment (seanbudd) [reply source/gui/__init__.py]: Hmm, sadly this can't be done for the destruction of the `wx.Menu` from `sysTrayIcon` as `sysTrayIcon` doesn't have a destruction event handler because it's not a window. I hope that also means that `wx` also won't be destroying it automatically.

I've fixed this for this function though.

Comment (feerrenrut) [reply source/gui/__init__.py]: Ah yep, that makes sense. It's a shame there are so many exceptions to learn about with wx!

File: source/gui/__init__.py [hunk 3]
```diff
@@ -81,7 +81,6 @@ def __init__(self):
 				self.Hide()
 
 	def _onDestroy(self, evt):
-		evt.Skip()  # Allow other handlers to process this event.
```
Comment (feerrenrut) [inline source/gui/__init__.py]: Why should `evt.Skip()` be removed? There shouldn't be any other parent controls, but if there were?

**Review by feerrenrut — COMMENTED** [review]
Could you also add a note into the "safe exit" function you added previously to point out that wx objects that don't inherit from wx.Window (eg sysTrayIcon, Menu) need to be manually destroyed.  That function is a logical place to look for an explanation if NVDA has problems during exit in the future.

File: source/gui/__init__.py (hunk 3 above)
Comment (seanbudd) [reply source/gui/__init__.py]: The install process was crashing, so I moved to safer behaviour of stopping and resuming the event propagation.

Comment (seanbudd) [reply source/gui/__init__.py]: I've added a comment that makes the reason for this more explicit

File: source/gui/__init__.py [hunk 4]
```diff
@@ -75,6 +80,13 @@ def __init__(self):
 				self.Show()
 				self.Hide()
 
+	def _onDestroy(self, evt: wx.EVT_WINDOW_DESTROY):
+		self.Skip(skip=False)  # we should block the destruction events until the sysTrayIcon is destroyed
+		# wx destroys child Windows automatically but `wx.adv.TaskBarIcon` is not a window
+		# so it must be set to be destroyed when destroying our main frame window (#12243)
+		log.debug(f"destroying systray icon")
+		wx.CallAfter(self.sysTrayIcon.Destroy)
```
Comment (feerrenrut) [inline source/gui/__init__.py:88]: Why is callAfter necessary here?

File: source/gui/__init__.py [hunk 5]
```diff
@@ -75,6 +80,13 @@ def __init__(self):
 				self.Show()
 				self.Hide()
 
+	def _onDestroy(self, evt: wx.EVT_WINDOW_DESTROY):
+		self.Skip(skip=False)  # we should block the destruction events until the sysTrayIcon is destroyed
```
Comment (feerrenrut) [inline source/gui/__init__.py:84]: "until" makes it sound like this logic will change at some point.

Hopefully I'm not alone in finding `skip` hard to memorize, I wish they had named it in the affirmative eg `handled`. 

The docs say:
>This method can be used inside an event handler to control whether further event handlers bound to this event will be called after the current one returns.
>
>Without Skip (or equivalently if Skip(false) is used), the event will not be processed any more. If Skip(true) is called, the event processing system continues searching for a further handler function for this event, even though it has been processed already in the current handler.
>
>In general, it is recommended to skip all non-command events to allow the default handling to take place. The command events are, however, normally not skipped as usually a single command such as a button click or menu item selection must only be processed by one handler.

This sounds like it will do the opposite of what the comment here says. 

**I've just noticed that this calls `self.Skip` rather than `evt.Skip` I assume this is a mistake?**

**Review by feerrenrut — CHANGES_REQUESTED** [review]

File: source/gui/__init__.py (hunk 5 above)
Comment (seanbudd) [reply source/gui/__init__.py:84]: I did find this hard to parse. I'm not sure what the default handler does for a `wx.EVT_WINDOW_DESTROY` event, but this is a command event so this behaviour is recommended and crashes otherwise. This event fires just before `self.Destroy` - which is what the comment was trying to allude to in a misleading way. I think the comment where the event handler is being bound should makes this clear anyway. 

And yes, the self/evt typo was a mistake.

File: source/gui/__init__.py (hunk 4 above)
Comment (seanbudd) [reply source/gui/__init__.py]: I can't find this in any official documentation, but when doing the original `safeAppExit` PR I found several recommendations in forums to use this method as it queues destruction events safer. Calling `Destroy` directly here causes a crash on the installation process.

File: source/gui/__init__.py [hunk 6]
```diff
@@ -75,6 +80,13 @@ def __init__(self):
 				self.Show()
 				self.Hide()
 
+	def _onDestroy(self, evt: wx.EVT_WINDOW_DESTROY):
+		evt.Skip(skip=False)  # blocks other handlers as wxCommandEvents must only be processed by one handler
+		# wx destroys child Windows automatically but `wx.adv.TaskBarIcon` is not a window
+		# so it must be set to be destroyed when destroying our main frame window (#12243)
+		log.debug(f"destroying systray icon")
```
Comment (lukaszgo1) [inline source/gui/__init__.py:87]: Why  you're using f-string here?

File: source/gui/__init__.py [hunk 7]
```diff
@@ -358,30 +370,35 @@ def onConfigProfilesCommand(self, evt):
...
 	"""
 
 	for window in wx.GetTopLevelWindows():
 		if isinstance(window, wx.Dialog) and window.IsModal():
-			log.info(f"ending modal {window} during exit process")
+			log.debug(f"ending modal {window} during exit process")
 			wx.CallAfter(window.EndModal, wx.ID_CLOSE_ALL)
 		if isinstance(window, MainFrame):
-			log.info(f"destroying main frame during exit process")
+			log.debug(f"destroying main frame during exit process")
```
Comment (lukaszgo1) [inline source/gui/__init__.py:382]: Same here w.r.t f-string being unnecessary.

File: source/gui/__init__.py [hunk 8]
```diff
@@ -542,6 +559,13 @@ def __init__(self, frame):
 		self.Bind(wx.adv.EVT_TASKBAR_LEFT_DOWN, self.onActivate)
 		self.Bind(wx.adv.EVT_TASKBAR_RIGHT_DOWN, self.onActivate)
 
+	def Destroy(self):
+		# wx.Windows destroy child Windows automatically but wx.Menu and TaskBarIcon don't inherit from
+		# wx.Window. The menu must be manually destroyed when destroying our system tray icon (#12243)
+		log.debug(f"destroying systray menu during exit process")
```
Comment (lukaszgo1) [inline source/gui/__init__.py:565]: And one more

**Comment by seanbudd** [comment]
closing as it appears this can't be fixed without inducing a crash
